node/core/dispute-coordinator/src/lib.rs

```diff
@@ -0,0 +1,959 @@
+	vec: &mut Vec<(T, ValidatorIndex, ValidatorSignature)>,
+	tag: T,
+	val_index: ValidatorIndex,
+	val_signature: ValidatorSignature,
+) {
+	let pos = match vec.binary_search_by_key(&val_index, |x| x.1) {
+		Ok(_) => return, // no duplicates needed.
+		Err(p) => p,
+	};
+
+	vec.insert(pos, (tag, val_index, val_signature));
+}
+
+fn handle_import_statements(
+	store: &dyn KeyValueDB,
+	state: &mut State,
+	config: &Config,
+	candidate_hash: CandidateHash,
+	candidate_receipt: CandidateReceipt,
+	session: SessionIndex,
+	statements: Vec<(SignedDisputeStatement, ValidatorIndex)>,
+) -> Result<(), Error> {
+	if state.highest_session.map_or(true, |h| session + DISPUTE_WINDOW < h) {
+		return Ok(());
+	}
+
+	let validators = match state.rolling_session_window.session_info(session) {
+		None => {
+			tracing::warn!(
+				target: LOG_TARGET,
+				session,
+				"Missing info for session which has an active dispute",
+			);
+
+			return Ok(())
+		}
+		Some(info) => info.validators.clone(),
+	};
+
+	let n_validators = validators.len();
+
+	let supermajority_threshold = polkadot_primitives::v1::supermajority_threshold(n_validators);
+
+	let mut votes = db::v1::load_candidate_votes(
+		store,
+		&config.column_config(),
+		session,
+		&candidate_hash
+	)?
+		.map(CandidateVotes::from)
+		.unwrap_or_else(|| CandidateVotes {
+			candidate_receipt: candidate_receipt.clone(),
+			valid: Vec::new(),
+			invalid: Vec::new(),
+		});
+
+	let was_undisputed = votes.valid.len() == 0 || votes.invalid.len() == 0;
+
+	// Update candidate votes.
+	for (statement, val_index) in statements {
+		if validators.get(val_index.0 as usize)
+			.map_or(true, |v| v != statement.validator_public())
+		{
+			tracing::debug!(
+				target: LOG_TARGET,
+				?val_index,
+				session,
+				claimed_key = ?statement.validator_public(),
+				"Validator index doesn't match claimed key",
+			);
+
+			continue
+		}
+
+		match statement.statement().clone() {
+			DisputeStatement::Valid(valid_kind) => {
+				insert_into_statement_vec(
+					&mut votes.valid,
+					valid_kind,
+					val_index,
+					statement.validator_signature().clone(),
+				);
+			}
+			DisputeStatement::Invalid(invalid_kind) => {
+				insert_into_statement_vec(
+					&mut votes.invalid,
+					invalid_kind,
+					val_index,
+					statement.validator_signature().clone(),
+				);
+			}
+		}
+	}
+
+	// Check if newly disputed.
+	let is_disputed = !votes.valid.is_empty() && !votes.invalid.is_empty();
+	let freshly_disputed = is_disputed && was_undisputed;
+	let already_disputed = is_disputed && !was_undisputed;
+	let concluded_valid = votes.valid.len() >= supermajority_threshold;
+
+	let mut tx = db::v1::Transaction::default();
+
+	if freshly_disputed && !concluded_valid {
+		// add to active disputes and begin local participation.
+		update_active_disputes(
+			store,
+			config,
+			&mut tx,
+			|active| active.insert(session, candidate_hash),
+		)?;
+	}
+
+	if concluded_valid && already_disputed {
```

Comment:
The `&& already_disputed` is only there for performance reasons I assume?

Reply:
I don't think it's possible for  `concluded_valid && freshly_disputed` to be `true`

`let concluded_valid = votes.valid.len() >= supermajority_threshold;`
`let was_undisputed = votes.valid.len() == 0 || votes.invalid.len() == 0;`

So `was_undisputed` => `!concluded_valid`

Therefore, isn't 

`conclude_valid && already_disputed` <=> `concluded_valid && is_disputed`

Reply:
I believe that due to the definition of 

`let was_undisputed = votes.valid.is_empty() || votes.invalid.is_empty();`

It holds that:

`concluded_valid && already_disputed` <=> `concluded_valid && is_disputed && !votes.invalid.is_empty();`

Reply:
`concluded_valid` is calculated after we update the votes, but `already_disputed` -- before, so
`was_undisputed` does not imply `!concluded_valid`

Reply:
Yes, took me a second to realize that. Thank you, removed my comment.

Reply:
Yep